generator/src/main/java/com/scottlogic/deg/generator/walker/ReductiveDecisionTreeWalker.java

```diff
@@ -1,107 +1,98 @@
 package com.scottlogic.deg.generator.walker;
 
 import com.google.inject.Inject;
+import com.scottlogic.deg.generator.Field;
 import com.scottlogic.deg.generator.FlatMappingSpliterator;
 import com.scottlogic.deg.generator.decisiontree.ConstraintNode;
 import com.scottlogic.deg.generator.decisiontree.DecisionTree;
+import com.scottlogic.deg.generator.fieldspecs.FieldSpec;
 import com.scottlogic.deg.generator.fieldspecs.RowSpec;
+import com.scottlogic.deg.generator.generation.FieldSpecValueGenerator;
 import com.scottlogic.deg.generator.generation.ReductiveDataGeneratorMonitor;
 import com.scottlogic.deg.generator.walker.reductive.*;
+import com.scottlogic.deg.generator.walker.reductive.fieldselectionstrategy.FieldValue;
 import com.scottlogic.deg.generator.walker.reductive.fieldselectionstrategy.FixFieldStrategy;
 
 import java.io.IOException;
-import java.util.Iterator;
-import java.util.Spliterator;
-import java.util.Spliterators;
+import java.util.Optional;
 import java.util.stream.Stream;
-import java.util.stream.StreamSupport;
 
 public class ReductiveDecisionTreeWalker implements DecisionTreeWalker {
-    private final ReductiveTreePruner treeReducer;
+    private final ReductiveTreePruner treePruner;
     private final IterationVisualiser iterationVisualiser;
-    private final FixedFieldBuilder fixedFieldBuilder;
+    private final ReductiveFieldSpecBuilder reductiveFieldSpecBuilder;
     private final ReductiveDataGeneratorMonitor monitor;
     private final ReductiveRowSpecGenerator reductiveRowSpecGenerator;
+    private final FieldSpecValueGenerator fieldSpecValueGenerator;
 
     @Inject
     public ReductiveDecisionTreeWalker(
         IterationVisualiser iterationVisualiser,
-        FixedFieldBuilder fixedFieldBuilder,
+        ReductiveFieldSpecBuilder reductiveFieldSpecBuilder,
         ReductiveDataGeneratorMonitor monitor,
-        ReductiveTreePruner treeReducer,
-        ReductiveRowSpecGenerator reductiveRowSpecGenerator) {
+        ReductiveTreePruner treePruner,
+        ReductiveRowSpecGenerator reductiveRowSpecGenerator,
+        FieldSpecValueGenerator fieldSpecValueGenerator) {
         this.iterationVisualiser = iterationVisualiser;
-        this.fixedFieldBuilder = fixedFieldBuilder;
+        this.reductiveFieldSpecBuilder = reductiveFieldSpecBuilder;
         this.monitor = monitor;
-        this.treeReducer = treeReducer;
+        this.treePruner = treePruner;
         this.reductiveRowSpecGenerator = reductiveRowSpecGenerator;
+        this.fieldSpecValueGenerator = fieldSpecValueGenerator;
     }
 
     /* initialise the walker with a set (ReductiveState) of unfixed fields */
     public Stream<RowSpec> walk(DecisionTree tree, FixFieldStrategy fixFieldStrategy) {
-        ConstraintNode rootNode = tree.getRootNode();
         ReductiveState initialState = new ReductiveState(tree.fields);
+        visualise(tree.getRootNode(), initialState);
```

Comment:
Should visualise always be called?

Reply:
it visualises for every stage of the tree reduction.
visualisation is disabled by default